### PR TITLE
Feature/externals v5

### PR DESCRIPTION
**Pull request by montegoulding**
Resolves some issues I found with java parameter handling.

### CODE REVIEW

File: lcidlc/src/InterfaceGenerate.cpp [hunk 1]
```diff
@@ -496,7 +496,7 @@ class JavaStringTypeMapper: public TypeMapper
 	
 	virtual void Default(CoderRef p_coder, ParameterType p_mode, const char *p_name, ValueRef p_value)
 	{
-		CoderWriteStatement(p_coder, "%s = default__java_string(__java_env, \"%s\", %s)", StringGetCStringPtr(p_value), p_name);
+		CoderWriteStatement(p_coder, "success = java_from__cstring(__java_env, \"%s\", %s) == kLCErrorNone", StringGetCStringPtr(p_value), p_name);
```
Comment (runrevmark) [inline lcidlc/src/InterfaceGenerate.cpp]: This case should be:

```
success = default__java_string(__java_env, \"%s\", %s);
```

Where default__java_string is a android-only method in support.mm:

```
static bool default__java_string(JNIEnv *env, const char *p_value, jobject& r_java_value)
{
    if (java_from__cstring(env, p_value, r_java_value))
      return true;
    return error__out_of_memory();
}
```

This ensures an appropriate error is thrown like the other default methods.

File: lcidlc/src/InterfaceGenerate.cpp [hunk 2]
```diff
@@ -1301,8 +1301,14 @@ static bool InterfaceGenerateVariant(InterfaceRef self, CoderRef p_coder, Handle
 			CoderWrite(p_coder, "returnvalue = __java_env -> CallStatic%sMethod(s_java_class, s_method", t_java_method_type);
 		else
 			CoderWrite(p_coder, "__java_env -> CallStaticVoidMethod(s_java_class, s_method");
-		for(uindex_t i = 0; i < p_variant -> parameter_count; i++)
-			CoderWrite(p_coder, ", t_param_%d", i);
+		
+        for(uint32_t k = 0; k < p_variant -> parameter_count; k++)
```
Comment (runrevmark) [inline lcidlc/src/InterfaceGenerate.cpp:1299]: Good catch - I think I made a mistake elsewhere too... The 'MappedParameter' list (t_mapped_params) is processed at the start with the name of the param var, so this should probably be:

```
for(uint32_t k = 0; k < p_variant -> parameter_count; k++)
    CodeWrite(p_coder, ", %s", t_mapped_params[k] -> var_name);
```

Similar when generating the native code parameter list.

File: lcidlc/src/Support.mm [hunk 3]
```diff
@@ -2923,31 +2965,13 @@ static bool fetch__java_string(JNIEnv *env, const char *arg, MCVariableRef var,
 	t_cstring_value = nil;
 	if (err == kLCErrorNone)
 		err = LCValueFetch(var, kLCValueOptionAsCString, &t_cstring_value);
+    
+    jobject t_java_value;
+    t_java_value = nil;
+    if (err == kLCErrorNone)
+        err = java_from__cstring(env, t_cstring_value, t_java_value);
 
-	size_t t_char_count;
```
Comment (runrevmark) [inline lcidlc/src/Support.mm:2927]: Clearly I didn't have my head screwed on quite right when writing this ;)

**Comment by runrevmark** [comment]
Hi Monte - thanks for catching these - have added a couple of comments above :)

File: lcidlc/src/InterfaceGenerate.cpp (hunk 1 above)
Comment (montegoulding) [inline lcidlc/src/InterfaceGenerate.cpp]: ok

File: lcidlc/src/InterfaceGenerate.cpp (hunk 2 above)
Comment (montegoulding) [inline lcidlc/src/InterfaceGenerate.cpp:1299]: ok

File: lcidlc/src/Support.mm (hunk 3 above)
Comment (montegoulding) [inline lcidlc/src/Support.mm:2927]: ;-)

**Comment by runrevmark** [comment]
HI Monte - thanks for this, it's now merged in.

**Pull request merged**